// @gardenfi/utils/dist/index24.js
import { hexToNumber as b } from "./index26.js";
import { formatLog as d } from "./index38.js";
import { transactionType as m } from "./index22.js";
const f = {
  "0x0": "reverted",
  "0x1": "success"
};
function r(l) {
  const u = {
    ...l,
    blockNumber: l.blockNumber ? BigInt(l.blockNumber) : null,
    contractAddress: l.contractAddress ? l.contractAddress : null,
    cumulativeGasUsed: l.cumulativeGasUsed ? BigInt(l.cumulativeGasUsed) : null,
    effectiveGasPrice: l.effectiveGasPrice ? BigInt(l.effectiveGasPrice) : null,
    gasUsed: l.gasUsed ? BigInt(l.gasUsed) : null,
    logs: l.logs ? l.logs.map((s) => d(s)) : null,
    to: l.to ? l.to : null,
    transactionIndex: l.transactionIndex ? b(l.transactionIndex) : null,
    status: l.status ? f[l.status] : null,
    type: l.type ? m[l.type] || l.type : null
  };
  return l.blobGasPrice && (u.blobGasPrice = BigInt(l.blobGasPrice)), l.blobGasUsed && (u.blobGasUsed = BigInt(l.blobGasUsed)), u;
}
export {
  r as formatTransactionReceipt,
  f as receiptStatuses
};
