// @gardenfi/react-hooks/dist/index3.js
import { filterDeadlineExpiredOrders as O, ParseOrderStatus as k } from "@gardenfi/core";
import { Orderbook as w } from "@gardenfi/orderbook";
import { useState as h, useEffect as l } from "react";
import { useWalletClient as v } from "wagmi";
const E = (o, n, m, s) => {
  const [r, p] = h(), { data: i } = v();
  return l(() => {
    if (!i || !o || !n) return;
    const e = new w({
      url: o,
      walletClient: i,
      auth: n
    });
    p(e);
  }, [i, o, n]), l(() => {
    !r || !s || s.fetchBlockNumbers().then((e) => {
      if (e.error) return;
      const { val: a } = e;
      r.fetchOrders(!0, !0, {
        per_page: 500
      }).then((u) => {
        if (u.error) return;
        const d = O(u.val.data).map((t) => {
          const c = a[t.source_swap.chain], f = a[t.destination_swap.chain];
          if (!c || !f) return;
          const b = k(
            t,
            c,
            f
          );
          return {
            ...t,
            status: b
          };
        }).filter((t) => t !== void 0);
        m(d);
      });
    });
  }, [r, s]), { orderbook: r };
};
export {
  E as useOrderbook
};
